### PR TITLE
VZ-3510 Move External DNS install to Go component

### PR DESCRIPTION
**Pull request by davisgomes**
# Description

This PR implements the install component for external DNS and the subsequent unit tests.

Fixes VZ-3510

# Checklist 

As the author of this PR, I have:

- [ ] Checked that I included or updated copyright and license notices in all files that I altered
- [ ] Added or updated unit tests for any new functions I added
- [ ] Added or updated integration tests if appropriate
- [ ] Added or updated acceptance tests if appropriate

Code reviewer, please confirm this PR:

- [ ] Addressed the requirement and meets the acceptance criteria
- [ ] Does not introduce unrelated or spurious changes
- [ ] Does not introduce any unapproved dependency
- [ ] Makes sense and it easy to understand, and/or difficult areas of code are clearly documented so that they can be understood

### CODE REVIEW

File: platform-operator/controllers/verrazzano/component/externaldns/external_dns.go [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+
+	// Get OCI DNS secret from the verrazzano-install namespace
+	dns := compContext.EffectiveCR().Spec.Components.DNS
+	dnsSecret := v1.Secret{}
+	if err := compContext.Client().Get(context.TODO(), client.ObjectKey{Name: dns.OCI.OCIConfigSecret, Namespace: constants.VerrazzanoInstallNamespace}, &dnsSecret); err != nil {
+		compContext.Log().Errorf("Could not find secret %s in the %s namespace: %s", dns.OCI.OCIConfigSecret, constants.VerrazzanoInstallNamespace, err)
+		return err
+	}
+
+	// Attach compartment to secret and apply it in the external DNS namespace
```
Comment (michael-cico) [inline platform-operator/controllers/verrazzano/component/externaldns/external_dns.go:44]: I think this comment means setting up all the OCI DNS secret info right?  Not just the compartment?

Comment (davisgomes) [reply platform-operator/controllers/verrazzano/component/externaldns/external_dns.go:44]: Yes, but the compartment field needs to be added to the original secret definition. That's why I chose to specify it. I get how that's a little confusing.

File: platform-operator/scripts/install/2-install-system-components.sh [hunk 2]
```diff
@@ -474,7 +429,7 @@ fi
 
 platform_operator_install_message "Installing cert-manager"
 if [ "$DNS_TYPE" == "oci" ]; then
-  action "Installing external DNS" install_external_dns || exit 1
+  platform_operator_install_message "Installing external DNS"
```
Comment (michael-cico) [inline platform-operator/scripts/install/2-install-system-components.sh:432]: Just a nit, for this message you can just give the component name, e.g., "External DNS".  the func spits out a message about the install having moved to the platform operator.

File: platform-operator/controllers/verrazzano/component/externaldns/external_dns.go (hunk 1 above)
Comment (michael-cico) [reply platform-operator/controllers/verrazzano/component/externaldns/external_dns.go:44]: l think I'm still confused, lol.  think I need to re-review how that stuff works.

File: platform-operator/scripts/install/2-install-system-components.sh (hunk 2 above)
Comment (davisgomes) [reply platform-operator/scripts/install/2-install-system-components.sh]: Thanks. Fixed!

**Review by michael-cico — DISMISSED** [review]

File: platform-operator/controllers/verrazzano/component/externaldns/external_dns.go [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+
+// ComponentName is the name of the component
+const (
+	ComponentName             = "external-dns"
+	externalDNSNamespace      = "cert-manager"
+	externalDNSDeploymentName = "external-dns"
+	ociSecretFileName         = "oci.yaml"
+)
+
+func (e externalDNSComponent) PreInstall(compContext spi.ComponentContext) error {
```
Comment (pfmackin) [inline platform-operator/controllers/verrazzano/component/externaldns/external_dns.go:29]: can you move the SPI functions to external_dns_component.go and they can just call the functions here (lowercase postInstall).  We are trying to have every SPI function in the component go file and keep that file small.  In this case you could just move the functions as is to that file there

File: platform-operator/controllers/verrazzano/component/externaldns/external_dns.go [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+	// Verify that the oci secret has one value
+	if len(dnsSecret.Data) != 1 {
+		compContext.Log().Errorf("OCI secret for OCI DNS should be created from one file")
+	}
+
+	// Extract data and create secret in the external DNS namespace
+	for k := range dnsSecret.Data {
+		externalDNSSecret.Data[ociSecretFileName] = append(dnsSecret.Data[k], []byte(fmt.Sprintf("  compartment: %s", dns.OCI.DNSZoneCompartmentOCID))...)
+	}
+	externalDNSSecret.Namespace = externalDNSNamespace
```
Comment (pfmackin) [inline platform-operator/controllers/verrazzano/component/externaldns/external_dns.go:58]: createOrUpdate should mutate in the callback func only.  Just the name and namespace needs to be populated in the secret before making the call

File: platform-operator/controllers/verrazzano/component/externaldns/external_dns.go (hunk 3 above)
Comment (davisgomes) [reply platform-operator/controllers/verrazzano/component/externaldns/external_dns.go:29]: Yea I can do that.

**Review by anders-swanson — DISMISSED** [review]
lgtm

**Review by michael-cico — DISMISSED** [review]

File: platform-operator/controllers/verrazzano/component/externaldns/external_dns_component.go [hunk 5]
```diff
@@ -13,14 +15,33 @@ type externalDNSComponent struct {
...
+			SupportsOperatorInstall: true,
+			ImagePullSecretKeyname:  secret.DefaultImagePullSecretKeyName,
 			ValuesFile:              filepath.Join(config.GetHelmOverridesDir(), "external-dns-values.yaml"),
+			AppendOverridesFunc:     AppendOverrides,
+			MinVerrazzanoVersion:    constants.VerrazzanoVersion1_0_0,
 		},
 	}
 }
+
+func (e externalDNSComponent) PreInstall(compContext spi.ComponentContext) error {
```
Comment (mbenmclean) [inline platform-operator/controllers/verrazzano/component/externaldns/external_dns_component.go:37]: Is this necessary to have the method call a helper function? I don't know if we do this for other components.

Comment (davisgomes) [reply platform-operator/controllers/verrazzano/component/externaldns/external_dns_component.go:37]: I don't think it's strictly necessary, but this is how Paul wanted me to format the component.

Comment (davisgomes) [reply platform-operator/controllers/verrazzano/component/externaldns/external_dns_component.go:37]: https://github.com/verrazzano/verrazzano/pull/1993#discussion_r752471448

Comment (mbenmclean) [reply platform-operator/controllers/verrazzano/component/externaldns/external_dns_component.go:37]: Ok that's fine just wondering this ins't the pattern I've seen in other components

**Review by mbenmclean — APPROVED** [review]

File: platform-operator/controllers/verrazzano/component/externaldns/external_dns_component.go (hunk 5 above)
Comment (michael-cico) [reply platform-operator/controllers/verrazzano/component/externaldns/external_dns_component.go:37]: It's not required.  I think we can revisit at some point.  Typically I've used helpers where I need to do the same thing between install/upgrade scenarios.  I think at some point some of the install/upgrade lifecycle operations will collapse into one, if it's possible.

**Pull request merged**